src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor_code.go

```diff
@@ -201,6 +201,63 @@ func (a *StoragePowerActorCode_I) ReportConsensusFault(rt Runtime, slasherAddr a
 	// include ReportUncommittedPowerFault(cheaterAddr addr.Address, numSectors util.UVarint) as case
 	// Quite a bit more straightforward since only called by the cron actor (ie publicly verified)
 	// slash cheater pledge collateral accordingly based on num sectors faulted
+
+	// validation checks
+	// - there should be exactly two blocks in proof
+	if len(proof) != 2 {
+		rt.AbortArgMsg("spa.ReportConsensusFault: proof requires two blocks")
+	}
+	// - both blocks are mined by the same miner
+	if proof[0].Miner() != proof[1].Miner() {
+		rt.AbortArgMsg("spa.ReportConsensusFault: proof needs to come from same miner")
+	}
+	// - first block is of the same or lower block height as the second block
+	if proof[0].Epoch() > proof[1].Epoch() {
+		rt.AbortArgMsg("spa.ReportConsensusFault: first block is of a higher block height")
+	}
+
+	// return from EC's IsValidConsensusFault
+	isValidConsensusFault := false
+
+	if !isValidConsensusFault {
+		rt.AbortArgMsg("spa.ReportConsensusFault: not valid consensus fault")
+	} else {
+		slasherAddr := rt.ImmediateCaller()
+		h, st := a.State(rt)
+		minerToSlash := proof[0].Miner()
+
+		claimedPower, powerOk := st.ClaimedPower()[minerToSlash]
+		if !powerOk {
+			rt.AbortArgMsg("spa.ReportConsensusFault: miner to slash has been slashed")
+		}
+		Assert(claimedPower > 0)
+
+		currPledge, pledgeOk := st._getCurrPledgeForMiner(minerToSlash)
+		if !pledgeOk {
+			rt.AbortArgMsg("spa.ReportConsensusFault: miner to slash has no pledge")
+		}
+		Assert(currPledge > 0)
+
+		// elapsed epoch from the latter block which committed the fault
+		elapsedEpoch := rt.CurrEpoch() - proof[1].Epoch()
+		if elapsedEpoch <= 0 {
+			rt.AbortArgMsg("spa.ReportConsensusFault: invalid block")
+		}
+
+		collateralToSlash := st._getPledgeSlashForConsensusFault(currPledge, faultType)
+		slasherShare := _getConsensusFaultSlasherShare(elapsedEpoch)
+		Assert(slasherShare <= 1 && slasherShare > 0)
+		amountToSlasher := actor.TokenAmount(slasherShare) * collateralToSlash
```

Comment:
Blocking: `TokenAmount` is an integer so the cast will clamp to either 0 or 1, which isn't what we want. 

I have a strong preference for reworking this arithmetic in terms of bigints. Rather than _getConsensusFaultSlasherShare returning a float, have it accept `collateralToSlash` as a parameter and compute the share using integer multiplication and division. Perhaps we can represent the constants as numerator/denominator ratios.

Reply:
I'm not sure on how to do BigInt arithmetic correctly atm, so I have updated the logic and left them in comment form.